### DISCUSSION TITLE
How to have childs component in "Smart Form Component"

### DISCUSSION THREAD

**Discussion by hersentino** [react-hook-form/react-hook-form#6904] [Q&A]
Hi, I'm facing a problem with the usage of "Smart Form Component" (https://react-hook-form.com/advanced-usage/#SmartFormComponent)

form.tsx (the same as the official doc)
```
 export default function Form({ defaultValues, children, onSubmit }) {
  const methods = useForm({ defaultValues });
  const { handleSubmit } = methods;

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {React.Children.map(children, child => {
        return React.createElement(child.type, {
              ...{
                ...child.props,
                register: methods.register,
                key: child.props.name
              }
            })
       })}
    </form>
  );
}

export function Input({ register, name, ...rest }) {
  return <input {...register(name)} {...rest} />;
}

```

this works absolutly well :

```
import Form, { Input } from './form'

const Parent = () => {
  return (
    <Form onSubmit={e => console.log(e)}>
      <Input name='email' />
      <button type='submit'>submit</button>
    </Form>
  )
}

```

this doesn't work :

```
import Form, {Input} from './form'

const Child = () => {
  return (<Input name='email' />)
}

const Parent = () => {
return (
   <Form onSubmit={e => console.log(e)}>
      <Child />
      <button type='submit'>submit</button>
    </Form>
  )
}
```

I'm getting this error :
<details>
  <summary>see the error</summary>
  
```
App.js:28 Uncaught TypeError: register is not a function
    at Input (App.js:28)
    at renderWithHooks (react-dom.development.js:14985)
    at mountIndeterminateComponent (react-dom.development.js:17811)
    at beginWork (react-dom.development.js:19049)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23964)
    at performUnitOfWork (react-dom.development.js:22776)
    at workLoopSync (react-dom.development.js:22707)
    at renderRootSync (react-dom.development.js:22670)
    at performSyncWorkOnRoot (react-dom.development.js:22293)
    at react-dom.development.js:11327
    at unstable_runWithPriority (scheduler.development.js:468)
    at runWithPriority$1 (react-dom.development.js:11276)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11322)
    at flushSyncCallbackQueue (react-dom.development.js:11309)
    at flushSync (react-dom.development.js:22467)
    at scheduleRefresh (react-dom.development.js:24429)
    at Object.renderer.scheduleRefresh (react_devtools_backend.js:5344)
    at react-refresh-runtime.development.js:304
    at Set.forEach (<anonymous>)
    at Object.performReactRefresh (react-refresh-runtime.development.js:293)
    at RefreshUtils.js:62
```
</details>

I suspect the React.Children.map of Form can't access to the childs of Child component. Does someone already experience this issue and have a solution ?

Thanks !

**Comment by lundn** [comment]
Maybe this line from the [React docs](https://reactjs.org/docs/react-api.html#reactchildren) will point you in the right direction:
> If children is a Fragment it will be treated as a single child and not traversed.

I'm not able to provide you with a solution or alternative setup atm. I guess you would need to pass the form methods to your Child component, and inside that one map the children and to the registration.

**Reply by hersentino** [reply]
Sorry about that, I did a mistake in my code. Fragment is not necessary in my code, I edited my post. Even without Fragment, I've the issue.
To reproduce, here is a GitHub repository with the example: https://github.com/hersentino/question-react-hook-form

**Comment by jfreedman0212** [comment]
When mapping over the children, there's a condition in the Form component that checks if it has a name prop. If it does (like in the first example), then it registers it to your form. However, if it doesn't (like in the second example), it won't work. The component won't go deeper into the children other than the first level (it will only look at Child and not the Input it renders).

EDIT: sorry, I deleted the message in the thread and put it out here, where I meant to put it

**Comment by hersentino** [comment]
Thanks for your answer ! 

Sorry I made a mistake, the code that I put in the first message was not update (the name condition should not be there). You can find a fully reproducible example without the "name condition" here: https://github.com/hersentino/question-react-hook-form. I have edited the first message.

**Comment by jfreedman0212** [comment]
@hersentino thanks for updating the example! Even without the `name` condition, I don't think that `React.Children.map` is going to go deeper than one level when rendering children. So, if you have an `Input` that isn't a _direct_ child of `Form`, then you won't be able to hook it into your `useForm`'s state.

I can think of a couple ways to get around this:
1. Pass the methods returned by `useForm` down into `Child` (I believe this is what @lundn was suggesting)
2. Use `FormProvider` and pass those values down via the Context API

If I have some time soon, I'll create some Code Sandboxes illustrating what I mean

**Reply by hersentino** [reply]
Thanks for your answer !

I guess FormProvider ([FormProviderPerformance](https://react-hook-form.com/advanced-usage#FormProviderPerformance)) could solve my problem ! I'll try soon and keep this discussions updated. But I have to give the props manually to the Children component. I would prefer not to give manually the props to the children component. If I don't find solution I'll do it manually but in all case I will keep the discussion updated
